matrix_client/api.py

```diff
@@ -836,3 +836,76 @@ def delete_devices(self, auth_body, devices):
             "devices": devices
         }
         return self._send("POST", "/delete_devices", content=content)
+
+    def upload_keys(self, device_keys={}, one_time_keys={}):
+        """Publishes end-to-end encryption keys for the device.
+
+        Said device must be the one used when logging in.
+
+        Args:
+            device_keys (dict): Optional. Identity keys for the device.
+            one_time_keys (dict): Optional. One-time public keys.
+        """
+        content = {}
+        if device_keys:
+            content["device_keys"] = device_keys
+        if one_time_keys:
+            content["one_time_keys"] = one_time_keys
+        return self._send("POST", "/keys/upload", content=content)
+
+    def query_user_keys(self, user_id):
+        """Query HS for public keys of all the given user devices.
+
+        Args:
+            user_id (str): The ID of the user whose devices keys to retrieve.
+        """
+        return self.query_keys({user_id: []})
+
+    def query_keys(self, user_devices, timeout=None):
+        """Query HS for public keys by user and optionally device.
+
+        Args:
+            user_devices (dict): The devices whose keys to download. Should be
```

Comment:
Please format these for rst as seen in other method docstrings where args need multi-line explanations.